README.md
### Javascript/Typescript SDK for Santander Digital Trust

This is a plain JS/TS library, it doesn't rely on any specific framework. <br>
As such, it shouldn't have any compatibility issues with other frameworks.

#### How to use it

Import the dependency as a dependency
```
npm install @gruposantander/rp-client-typescript
```

The first step is to create the client and set it up:

```js
const verifyidclient = await VerifiedIdClient.createInstance({
  wellKnownURI: 'https://op-iamid-verifiedid-pro.e4ff.pro-eu-west-1.openshiftapps.com/.well-known/openid-configuration',
  privateJWK: resolve('./secrets/privateKey.json'),
  clientId: '12345678-aaaa-bbbb-cccc-1234567890ab'
}) // creating the client
await verifyidclient.setUpClient() // running all the setup steps (querying the well-known endpoint and storing the public jwks)
```

After the step above, it is possible to initiate the flow straight away, by preparing the request and then send it:

```js
const claims = new Claims()
claims.email().withEssential(true).withPurpose('email purpose')
claims.givenName().withIAL(2)
claims.lastYearMoneyIn().withPurpose('Last year money in purpose')

const assertionClaims = new AssertionClaims()
assertionClaims.age().gt(21).withPurpose('age purpose')
assertionClaims.address()
  .withAssertion(Address.postalCode().eq('MK1 1AA'))
  .withAssertion(Address.country().eq('UK'))
  .withPurpose('Address purpose')

assertionClaims.lastYearMoneyIn()
  .withPurpose('Last year money in purpose')
  .withAssertion(Balance.amount().gt(200))

const request = new InitiateAuthorizeRequestBuilder()
  .withRedirectURI('com.myApp://callback')
  .withAssertionClaims(assertionClaims)
  .withClaims(claims)
  .withPurpose('top level purpose')
  .build()
```

While `InitiateAuthorizeRequestBuilder` follows a builder pattern, `Claims` and `AssertionClaims` use a different 
syntax to help readability

After the claims have been set up, the _/initiate-authorize_ can be invoked:

```js
const initiateAuthorizeResponse = await verifyidclient.initiateAuthorize(request)
```

From the above reponse, `initiateAuthorizeResponse.redirectionUri` can be used to redirect users to the consent journey:
```
https://op-iamid-verifiedid-pro.e4ff.pro-eu-west-1.openshiftapps.com/web/login?request_uri=urn:op.iamid.io:JDAQ9YwMSLcCbiUz0Wq0HGjpu-wr4HngFDCv8tTkQa-
```
The consent process (happening in a browser, out of scope for this SDK) finishes with an authorization code:
```
302 com.myApp://callback?code=Ian06qtqg5PNUhfRUy9UFLwx4T7DIzodBLiFjoFYWmr
```
This code can then be used to invoke _/token_:

```js
const request = new TokenRequestBuilder()
  .withRedirectUri('com.myApp://callback')
  .withCode('Ian06qtqg5PNUhfRUy9UFLwx4T7DIzodBLiFjoFYWmr')
  .build()

const token = await verifyidclient.token(request)
```

## Supported Claims - Sharing
The following claims can be requested for sharing:
```js
const sharingClaims = new Claims()

sharingClaims.givenName()
sharingClaims.familyName()
sharingClaims.birthdate()
sharingClaims.gender()
sharingClaims.countryOfBirth()
sharingClaims.title()
sharingClaims.nationality()
sharingClaims.civilStatus()
sharingClaims.age()
sharingClaims.companyRegisteredName()
sharingClaims.companyTradeName()
sharingClaims.companyStartDate()
sharingClaims.companyEndDate()
sharingClaims.companyType()
sharingClaims.companyCountryIncorporation()
sharingClaims.companyAge()
sharingClaims.companyOperating()
sharingClaims.phoneNumber()
sharingClaims.email()
sharingClaims.address()
sharingClaims.totalBalance()
sharingClaims.lastYearMoneyIn()
sharingClaims.lastQuarterMoneyIn()
sharingClaims.averageMonthlyMoneyIn()
sharingClaims.passportId()
sharingClaims.drivingLicenseId()
sharingClaims.nationalCardId()
```

For each claim, a specific value _purpose_, _ial_ and/or _essential_ can be specified:
```js
sharingClaims.givenName().withPurpose("given name purpose").withIAL(3).withEssential(true)
sharingClaims.familyName().withIAL(TWO).withPurpose("family name purpose")
sharingClaims.birthdate().withEssential(false)
sharingClaims.gender()
```
None of these attributes are mandatory and they can be applied in any order

## Supported Claims - Verifying

The SDK supports the same verifying claims as it does for sharing, though the data model is quite different

There are three types of assertions in this SDK: _simple_, _comparative_ and _complex_.
 * *Simple assertions*:  they support only the ``.eq`` (equal assertion), e.g.
    ```js
    assertionClaims.email().eq("jane.doe@santander.co.uk")
    ```
   which translates into ``Verify that the email is "jane.doe@santander.co.uk"``
 * *Comparative assertions*: they extend _simple_ assertions with the ``.gt``, ``.gte``, ``.lt`` and ``.lte`` operations, e.g.:
   ```java
    assertionClaims.age().gte(18)
    ``` 
   which translates into ``Verify that the age is greater or equal than 18``
 * *Complex claims*: they include different properties, that can be individually asserted, for example:
    ```js
    assertionClaims.totalBalance()
        .withAssertion(Balance.currency().equal(Currency.getInstance("GBP")))
        .withAssertion(Balance.amount().gt(BigDecimal.valueOf(99.99)))
    ```
   which translates into ``Verify that the balance currency is equal to GBP and the amount is greater than 99.99``. 
   Depending on the individual claim, the property can either support only ``.eq`` or the entire set of comparisons 

Similarly to sharing claims, _purpose_, _ial_ and/or _essential_ can be attached to verifying claims:
```js
assertionClaims.address()
    .withAssertion(Address.country().eq("UK"))
    .withAssertion(Address.postalCode().eq("MK11AA"))
    .withPurpose("This is why RP is verifying your address")
```

To keep the API simple and concise, ``Balance`` and ``Address`` helpers have been created with the following methods:
```js
Balance.currency()
Balance.amount()
Address.formatted() 
Address.streetAddress()
Address.postalCode()
Address.locality()
Address.region()
Address.country()
```

## PKCE and nonce support
This SDK supports nonce verification and PKCE. Depending on the scenario one or both of them will be automatically enforced.
* Nonce: nonces can be passed as part of the ``InitiateAuthorizeRequest`` object:
   ```java
    const request = new InitiateAuthorizeRequestBuilder()
      .withRedirectURI('com.myApp://callback')
      .withAssertionClaims(assertionClaims)
      .withClaims(claims)
      .withPurpose('top level purpose')
      .withNonce('nonce-1111')
      .build()
    ```
    If such a value is not passed, the SDK will create one and will return it as part of the response object from the ``initiateAuthorize`` method.
    Regardless of whether the nonce is automatically or manually generated, such value will need to be passed when invoking the ``token`` method.
 * PKCE: similarly to nonce, a code challenge can be passed:
   ```java
   InitiateAuthorizeRequest request = InitiateAuthorizeRequest.builder()
       .redirectUri("https://example.com/callback")
       .claims(idClaims)
       .assertionClaims(assertionClaims)
       .codeChallenge("ii2ebegd173dg")
       .build()
   ```
   Different to the nonce case, a code verifier will be automatically generated (and the challenge calculated) if not passed to it *and* the redirectURI is a _deep link_. 
   If a challenge is passed or a verifier is generated, the verifier needs to be passed when invoking the ``token`` method:
   ```java
    TokenRequest tokenRequest = TokenRequest.builder()
        .redirectUri("com.myApp://callback")
        .authorizationCode("27IyhbWvL5uGY61f69A-RlEl7N2qRLm5vQ7_mO0tRGH")
        .codeVerifier("H-jwul2I2vbDb90ll-lfl14LXtES9lqZvgtiX3WYF44")
        .build()
    ```





CHANGELOG.md
# Changelog 
### 0.4.1
Added external publish details for github package

### 0.4.0
* Assertion language has removed all usages of `$`, now the syntax is:
  * eq 
  * gt 
  * gte 
  * lt
  * lte 
* Complex claims now contain a `props` field inside the claim, which contains the properties relating to the actual assertions being made. There are no changes in usage for a developer 

### 0.3.1
* Downgraded TypeScript to v3.5.3

#### 0.3.0
* Adding the following claims:
  * gender
  * country_of_birth
  * title
  * nationality
  * civil_status
  * age
  * company_registered_name
  * company_trade_name
  * company_start_date
  * company_end_date
  * company_type
  * company_country_incorporation
  * company_age
  * company_operating
  * last_year_money_in
  * last_quarter_money_in
  * average_monthly_money_in
  * passport_id
  * driving_license_id
  * national_card_id
* Renaming `.equal` to `.eq`

#### 0.2.1
* Fixing bug that didn't allow more than one sharing claim at a time

#### 0.2.0
* First MVP, support for the following claims (sharing & verifying):
  * phone_number
  * email
  * family_name
  * given_name
  * birthdate
  * age
  * total_balance
  * address



examples/example.js
const { Claims, AssertionClaims, Address, Balance } = require('@gruposantander/rp-client-typescript').Model
const { VerifiedIdClient, InitiateAuthorizeRequestBuilder, TokenRequestBuilder } = require('@gruposantander/rp-client-typescript').Client

const resolve = require('path').resolve

const wellKnownUri = 'https://op.iamid.io/.well-known/openid-configuration';

const claims = new Claims()
claims.email().withEssential(true).withPurpose('email purpose')
claims.givenName().withIAL(2)
claims.lastYearMoneyIn().withPurpose('Last year money in purpose')

const assertionClaims = new AssertionClaims()
assertionClaims.age().gt(21).withPurpose('age purpose')
assertionClaims.address()
  .withAssertion(Address.postalCode().eq('MK9 1BB'))
  .withAssertion(Address.country().eq('UK'))
  .withPurpose('Address purpose')

assertionClaims.lastYearMoneyIn().withPurpose('Last year money in purpose')
  .withAssertion(Balance.amount().gt(200))


// tokenExample()
doInitAuthorize()

async function doInitAuthorize() {
  try {
    const verifyidclient = await VerifiedIdClient.createInstance({
      wellKnownURI: wellKnownUri,
      privateJWK: resolve('./secrets/quick-jobs.json'),
      clientId: 'IIvRB-z9e0mTVDfrXpAsy'
    })
    await verifyidclient.setUpClient()
    const request = new InitiateAuthorizeRequestBuilder()
      .withRedirectURI('http://localhost:4201/profile')
      .withAssertionClaims(assertionClaims)
      .withClaims(claims)
      .withPurpose('top level purpose')
      .build()
    const initiateAuthorize = await verifyidclient.initiateAuthorize(request)
    console.log(`code verifier: ${initiateAuthorize.codeVerifier}`)
    console.log(`click here: ${initiateAuthorize.redirectionUri}`)
  } catch (e) {
    console.error(e)
  }
}

async function tokenExample() {
  try {
    const verifyidclient = await VerifiedIdClient.createInstance({
      wellKnownURI: wellKnownUri,
      privateJWK: resolve('./secrets/quick-jobs.json'),
      clientId: 'IIvRB-z9e0mTVDfrXpAsy'
    })
    await verifyidclient.setUpClient()
    const request = new TokenRequestBuilder()
      .withRedirectURI('http://localhost:4201/profile')
      .withCodeVerifier('gekBHnGA5NLuX2z-fVNIrdpbPAFfUj0hthCwYPXji7d')
      .withCode('07G60dVyAmkUxcqI7YXd5TPwx2nA_GXhfLHWRxS69UG')
      .build()
    const token = await verifyidclient.token(request)
    console.log('token', token)
  } catch (e) {
    console.error(e)
  }
}
